File: JAVA/Binary Tree Zigzag Level Order Traversal.java
public class Solution {
    public ArrayList<ArrayList<Integer>> zigzagLevelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> ans = new ArrayList<ArrayList<Integer>>();
        if (root == null) return ans;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);
        int lastLevel = 1;
        int thisLevel = 0;
        boolean direction = true;
         
        ArrayList<Integer> row = new ArrayList<Integer>();
        while (queue.size() != 0)
        {
            TreeNode n = queue.poll();
            lastLevel--;
            if (n.left != null)
            {
                queue.offer(n.left);
                thisLevel++;
            }
            if (n.right != null)
            {
                queue.offer(n.right);
                thisLevel++;
            }
            row.add(n.val);
            if (lastLevel == 0)
            {
                lastLevel = thisLevel;
                thisLevel = 0;
                if (!direction)
                {
                    // reverse
                    int len = row.size();
                    for (int i = 0; i < len/2; i++)
                    {
                        int tmp = row.get(i);
                        row.set(i, row.get(len-1-i));
                        row.set(len-1-i, tmp);
                    }
                }
                direction = !direction;
                ans.add(row);
                row = new ArrayList<Integer>();
            }
        }
        return ans;
    }
}

File: Binary Tree Zigzag Level Order Traversal.cpp
class Solution {
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
        vector<vector<int> > result;
        if (root == NULL) return result;
        queue<TreeNode*> que;
        que.push(root);
        int thisLevel = 1;
        int nextLevel = 0;
        vector<int> level;
        while (!que.empty()) {
            TreeNode *node = que.front();
            que.pop();
            level.push_back(node->val);
            thisLevel--;
            if (node->left != NULL) {
                que.push(node->left);
                nextLevel++;
            }
            if (node->right != NULL) {
                que.push(node->right);
                nextLevel++;
            }
            if (thisLevel == 0) {
                if (result.size() % 2 == 1) {
                    reverse(level.begin(), level.end());
                }
                result.push_back(level);
                level.clear();
                thisLevel = nextLevel;
                nextLevel = 0;
            }
        }
        return result;
    }
};
